File: src/block/BlockState.java
package block;

import world.Chunk;

public final class BlockState {
    private final Block block;

    public final byte[] properties;
    private final boolean isFullCube;
    private final boolean[] cullsFace;
    public final BlockModel model;

    public BlockState(Block block, byte... properties) {
        if (properties.length != block.attributes.length) throw new RuntimeException();
        this.block = block;
        this.properties = properties;
        this.model = block.getModel(this);
        this.cullsFace = new boolean[6];
        this.isFullCube = block.isFullCube(this);

        if (this.model == null) return;
        short[][][] faces = this.model.faces;
        for (int i = 0; i < faces.length; i++) {
            short[][] dir = faces[i];
            for (short[] face : dir) {
                if (face[0] == 0 && face[1] == 0 && face[2] == 0 && face[3] == 16 && face[4] == 16) {
                    this.cullsFace[i % 2 == 0 ? i + 1 : i - 1] = true;
                    break;
                }
            }
        }
    }

    public String toString() {
        StringBuilder str = new StringBuilder(this.block.name);
        str.append(":[");
        for (int j = 0; j < this.properties.length; j++) {
            int i = this.properties[j];
            str.append(this.block.attributes[j].name());
            str.append('=');
            if (this.block.attributes[j].property() instanceof BlockProperty.NamedBlockProperty) {
                str.append('\'');
                str.append(((BlockProperty.NamedBlockProperty) this.block.attributes[j].property()).getName(i));
                str.append('\'');
            }
            else str.append(i);
            if (j < this.properties.length - 1) str.append(',');
        }
        str.append("]");
        return str.toString();
    }

    public boolean isFullCube() {
        return this.isFullCube;
    }

    public boolean hasModel() {
        return this.model != null;
    }

    public boolean cullsFace(Chunk.Face face) {
        return this.cullsFace[face.ordinal()];
    }
}

File: java/src/block/BlockState.kt
package src.block

import src.block.BlockProperty.NamedBlockProperty

class BlockState(block: Block, vararg properties: Byte) {
    private val block: Block
    val properties: ByteArray
    val model: BlockModel
    val hasModel: Boolean

    init {
        if (properties.size != block.attributes.size) throw RuntimeException()
        this.block = block
        this.properties = properties
        this.hasModel = block.hasModel(this)
        model = block.getModel(this)
    }

    override fun toString(): String {
        val str = StringBuilder(block.name)
        str.append(":[")
        for (j in properties.indices) {
            val i = properties[j].toInt()
            str.append(block.attributes[j].name)
            str.append('=')
            if (block.attributes[j].property is NamedBlockProperty) {
                str.append('\'')
                str.append((block.attributes[j].property as NamedBlockProperty).getName(i))
                str.append('\'')
            } else str.append(i)
            if (j < properties.size - 1) str.append(',')
        }
        str.append("]")
        return str.toString()
    }
}